[[ TileCache ]] Use HardwareBuffer for direct memory access on Android

livecodeian opened the pull request:
This patch modifies the creation and update of GL textures to use a
native hardware buffer. This allows fast upload of texture data by
accessing the underlying buffer and copying pixel data to the texture
directly.

To do:
- [x] Make compatible with API < 26 (AHardwareBuffer calls need to be dlsym'd)
- [ ] Make work with 32-bit ARM

Of these the first is critical (otherwise the engine will not work on anything less than Android 8.0). The second would be nice, but isn't hugely important - 64-bit devices are common-place, even more so as API level increases.

livecodeian committed:
[[ TileCache ]] Disable HardwareBuffer on 32bit armv7 arch

This patch allows armv7 builds to complete successfully until the NDK
can be updated to one that includes the required headers
(android/hardware_buffer.h).

livecodeian committed:
[[ Tilecache ]] Use HardwareBuffer for direct memory access on Android

This patch modifies the creation and update of GL textures to use a
native hardware buffer. This allows fast upload of texture data by
accessing the underlying buffer and copying pixel data to the texture
directly.

livecodeian committed:
[[ TileCache ]] Add external texture program to MCGLContext

This patch adds an "external texture" program type to the glcontext
implementation which is similar to the regular texture program except
that an EGL extension is used to access an external texture.

runrevmark commented on engine/src/glcontext.cpp:
```diff
@@ -20,6 +20,12 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 #elif defined(TARGET_SUBPLATFORM_ANDROID)
+
+/* TODO - Don't attempt to build on 32bit armv7 - needs ndk update */
+#if !defined(__arm__)
```
So I need an idea on what (at least!) is involved in doing this to work out if it is worth the effort.

runrevmark commented on engine/src/glcontext.cpp:
```diff
@@ -258,6 +264,27 @@ void main()
 )glsl";
 
+#if defined(MCANDROIDUSEHARDWAREBUFFER)
+static const char *s_texture_fragment_shader_source =
```
I take it this shader program works when hardware buffers are not available too? (The reason I ask is that we seem to require the extension, and use samplerExternalOES in the specialized one).

livecodeian replied:
It doesn't - I've addressed this in an update to this PR so that now both the regular texture and external texture versions are available

runrevmark commented on engine/src/tilecachegl3.x.cpp:
```diff
@@ -181,18 +347,57 @@ static bool MCTileCacheOpenGLCompositorCreateTile(MCTileCacheOpenGLCompositorCon
+	t_have_hw_buffer = false;
+
+#if defined(MCANDROIDUSEHARDWAREBUFFER)
+	// create hardware buffer
```
I suggest refactoring this code so that all the hardware buffer specific stuff happens in its own method:
```
if (!MCTileCacheHardwareBufferCreate(kSuperTileSize, t_super_tile))
{
    ... do stuff for init not using hw buffers ...
}
```

livecodeian replied:
Fixed in updated commits

runrevmark commented on engine/src/tilecachegl3.x.cpp:
```diff
@@ -288,7 +522,12 @@ bool MCTileCacheOpenGLCompositor_EndTiling(void *p_context)
+#endif
+
 	// Flush any empty textures.
 	MCTileCacheOpenGLCompositorFlushSuperTiles(self, false);
```
It looks like we can elide HardwareBuffer unlock into Flush - i.e. just have:
```
MCTileCacheHardwareBufferUnlock(t_super_tile)
```
Called on each super-tile which exists.

livecodeian replied:
Fixed in updated commits

runrevmark commented on engine/src/tilecachegl3.x.cpp:
```diff
@@ -301,27 +540,80 @@ bool MCTileCacheOpenGLCompositor_AllocateTile(void *p_context, int32_t p_size, c
+	bool t_have_hw_buffer = false;
+
+#if defined(MCANDROIDUSEHARDWAREBUFFER)
+	t_have_hw_buffer = t_super_tile->hw_buffer != nil;
```
Similar to the above rather than use #ifdef here - instead make a higher-level method which fails if no hardware buffer. i.e.
if (!MCTileCacheHardwareBufferUpdate(super_tile, p_size, p_bits, p_stride, p_x, p_y))
{
   ... slow non-hardware buffer path ...
}
```

livecodeian replied:
Fixed in updated commits

runrevmark commented on engine/src/tilecachegl3.x.cpp:
```diff
@@ -347,17 +634,12 @@ bool MCTileCacheOpenGLCompositor_AllocateTile(void *p_context, int32_t p_size, c
 
-		// Set the tile id.
-		t_tile = (void *)t_tile_id;
+		if (t_data != p_bits)
```
Was this leaking before?!?

livecodeian replied:
No, this change is just there to move glTexSubImage2D and related code into an if (<no hardware buffer>) block

runrevmark reviewed (changes requested):
Overall I suggest making a simple API which returns false if hardwarebuffer stuff fails for the ciritcal places - perhaps MCTileCacheHardwareBuffer prefix. (Seems we only need Create / Unlock / Update / Destroy).

I suggest using forward references near the top, then implementing them at the bottom in an ifdef. If there is no hardwarebuffer implementation then the methods should just return false meaning the non-hardwarebuffer code is used.

This means that the only places we need to use the ifdefs are in the super_tile struct and around the methods - with dummy impls of the latter in the alternate.

Also, we need to make sure we dlsym the AHardwareBuffer symbols so things run on pre-8.0.

livecodeian committed:
[[ TileCache ]] Add GLHardwareBuffer API and Android implementation

This patch adds an API for obtaining a hardware buffer that can be bound
to a GL texture and locked for direct access to the memory buffer.

This is implemented on Android using the AHardwareBuffer C api. The API
functions are dynamically loaded to avoid linkage failures when
attempting to run on Android versions that don't support the APIs.

livecodeian commented:
@runrevmark I've revised this PR so that platform-specific code is moved to a seperate source file, and implemented the fallback for Android < 26. I've also abstracted the hardwarebuffer stuff generally and moved the #ifdef`d stuff to the bottom of the tilecache source as advised
